### ISSUE TITLE
Bundle many small jobs into one larger job submission

### ISSUE THREAD

**Issue by fischuu** [snakemake/snakemake#872]
**Is your feature request related to a problem? Please describe.**
Depending on the definition of the rules, it might happen that Snakemake creates huge amount of jobs submitted to a cluster

**Describe the solution you'd like**
In order to reduce the amount of jobs created by Snakemake, it would be really nice if Snakemake could check the resource allocations from different rules, combines them then into a single script (with more resources) and submits them together as one job. 

For example, if subgraphs of the DAG would have a combined runtime less than some parameter, these jobs would be submitted together as one script in right order with the accumulated resource allocations.

**Describe alternatives you've considered**
Currently I put commands with short runtime together into one rule, but it would make a workflow more readable if each task would have it's own rule.

**Comment by ASLeonard** [comment]
That sounds quite similar to the [job grouping feature](https://snakemake.readthedocs.io/en/stable/executing/grouping.html). It may not be as nuanced as you want, but it is a pretty straightforward way to bundle together many small jobs with some internal handling of pooled resources.

**Comment by fischuu** [comment]
Thanks, I was thinking first the same that the group feature could be used for it, but isn't it so that with the group-option one can only control which jobs will be submitted to the cluster at the same time? But they would still have their own resource allocations, slurm-jobid, etc, isn't it? I tried to find some examples for group, if it can be used also to merge rules into single jobs, but couldn't find any.

But maybe I misunderstood how group can be used, do you know if there is some webpage with more detailed use-cases of the job-grouping feature?

**Comment by ASLeonard** [comment]
From experience, if you run 10 tasks in the same group, setting `--group-components mygroup=5` would launch 2 unique jobs, `--group-components mygroup=10` would only launch a single unique job. The resources get combined, so if each task needs 1 core, the examples request 5 and 10 cores respectively. 

There is room for improvement here imo, as if I have 100 small jobs all with the same resource needs, I'd prefer `group` to just concatenate them, and still only request 1x job worth of resources.

**Comment by fischuu** [comment]
Great, thanks a lot, at least that is then something to play around and familiarize myself with it. I totally misunderstood this part of the manual and the use of `group`. Also, even if there is space for improvement, I think it will make a difference in my workflows, as I currently have 1.000+ samples with 30.000+ jobs, so if `group` could bring it down to 5.000-10.000 jobs that would be already something!

**Comment by jdblischak** [comment]
> There is room for improvement here imo, as if I have 100 small jobs all with the same resource needs, I'd prefer group to just concatenate them, and still only request 1x job worth of resources.

+1 to this. I have a single rule that has to be run tens of thousands of times and each runs for only a few seconds. I like how the single rule is very readable, but the bottleneck is the number of job submissions. I read the documentation on [job grouping](https://snakemake.readthedocs.io/en/stable/executing/grouping.html), and figured I could use something like the example to run a bunch of jobs in each submission. But when I requested 50 jobs per group, i.e. ` --groups somerule=group0 --group-components group0=50`, it requested 50 threads for this one job! Since the compute nodes on my cluster don't have 50 nodes, it can't be submitted.

**Comment by fischuu** [comment]
> There is room for improvement here imo,

I can agree with that as well, the group option has a huge potential, but is not very usable. At least after some test runs I came back to the point not using it, exactly because of the issues with the automatic resource allocations. 

Currently, I face the same issue - I concatenate files for hundreds of samples on a cluster with hundreds of nodes and this creates apparently a huge I/O pressure on the discs. So, it would be excellent if it would be possible to group and combine rules in a more flexible way and also adjust the resources accordingly.

**Comment by jdblischak** [comment]
Update: PR #1218 from @pvandyken was merged into main in commit a8014d030a2a3ea04743e30bcf5164801291378f and released in version 7.11.0 (#1783)

**Comment by jdblischak** [comment]
I just tested job grouping with Snakemake 7.11. Unfortunately, at least for my [intended use case](https://github.com/jdblischak/smk-simple-slurm/tree/main/examples/job-grouping), it still summed the `mem_mb` and `threads` for each of the rules in the group. So if the group job tries to include 100 instances of a rule, it requests 100 threads.

**Comment by pvandyken** [comment]
@jdblischak, did you specify the maximum number of threads using `--cores <N>` or `-c<N>`?

**Comment by jdblischak** [comment]
> did you specify the maximum number of threads using --cores <N> or -c<N>?

@pvandyken Thanks for the tip. I missed that new detail.

I tried out using `--cores` and `--resources mem_mb=x`. I have to admit that I find it difficult to reason about. When I use one or the other, both of the resources get restrained. For example, when I set `--cores 3`, each group job gets assigned 3 threads and 300 MB of memory. And if I set `--resources mem_mb=400`, then each group job gets assigned 4 threads and 400 MB of memory. And 100 rules still get bundled into a single group job.

Also, setting `--cores` to control the behavior of group jobs can affect other rules. I created another test rule that requests 8 threads, but it only uses whatever I set with `--cores`.

To summarize, your PR is an improvement and a step in the right direction (and I very much appreciate all the effort you put into it!), but I still find group jobs to be difficult to incorporate into a large Snakemake pipeline.

**Comment by pvandyken** [comment]
> I tried out using `--cores` and `--resources mem_mb=x`. I have to admit that I find it difficult to reason about. When I use one or the other, both of the resources get restrained. For example, when I set `--cores 3`, each group job gets assigned 3 threads and 300 MB of memory. And if I set `--resources mem_mb=400`, then each group job gets assigned 4 threads and 400 MB of memory. And 100 rules still get bundled into a single group job.

So it's not that both resources are being constrained, but that the scheduler requests the minimal amount of each resource needed. In your example, each job uses `100mb` of memory and 1 thread. When you constrain memory to `400mb`, the scheduler calculates that up to 4 jobs in the group can run simultaneously without going over the memory constraint. Because each job uses one core and only 4 are expected to run simultaneously, only 4 cores are requested. If it's important that more cores be used, either lower the memory requirement of the rule (or eliminate it), or increase the memory constraint (or eliminate it). In both cases of elimination, the default memory requested by slurm would need to be configured in the profile, or it will fall back to slurm defaults (1000mb)


> And 100 rules still get bundled into a single group job.

Correct, the new group resource calculations have no effect on `--connected-components`. In fact, the whole point was to allow large numbers of jobs to be bundled into one group while still respecting core and resource limitations.

> Also, setting --cores to control the behavior of group jobs can affect other rules. I created another test rule that requests 8 threads, but it only uses whatever I set with --cores.

Correct, the current implementation assumes that all nodes are the same. The logic is, for instance, if you set `--cores` as 4, all nodes have a maximum of 4 cores. So any jobs that are scheduled, group or not, will be limited to 4. But note that if one ran the workflow without a cluster and set `--cores 4`, all the jobs would still be limited to 4 cores, so the behaviour is consistent.

`--resources` and `--cores` is not meant to provide fine-grained control over the resource usage of each rule. For that, one has `--set-resources` and `--set-threads`. Rather, it's meant to provide the overall constraint for each node.

I fully realize most compute environments have a heterogeneous selection of nodes, with varying amounts of eg cores, memory, etc. A nice future addition might allow the setting of separate constraints for each type of node and assignment of jobs to different environments, but that would be another rather large undertaking. Although a certain version of this is semi-possible using local rules: one can use `--local-threads` to set a different thread limit for those rules annotated as `localrules: ...`

**Comment by jdblischak** [comment]
@pvandyken Thanks for the very thorough explanation. This helped me understand all the constraints on the implementation.

> `--resources` and `--cores` is not meant to provide fine-grained control over the resource usage of each rule. For that, one has `--set-resources` and `--set-threads`. Rather, it's meant to provide the overall constraint for each node.

This the fundamental disconnect for me. At the end of the day, I need to pass a number to Slurm's `sbatch` flag `--cpus-per-task`. I typically use `--cpus-per-task={threads}`. This breaks group jobs since the sum of the number of jobs is greater than the number of cores on any compute node. I tried to get around this limitation by specifying a fractional number of threads, but reasonably Snakemake didn't accept this trick:

```sh
set-threads:
  - standardRule=8
  - ruleToBeGrouped=0.01
```

Which gave me the error:

```python
Invalid threads definition: entries have to be defined as RULE=THREADS pairs (with THREADS being a positive integer).
```

I next tried to define my own resource to manage the per-rule threads, `mythreads`. Unfortunately it gets summed for grouped job just like `threads` and `mem_mb`, so it didn't fix the problem.

As far as I can tell, if you submit jobs to Slurm via `snakemake --cluster` like I do, and you need to specify multiple threads via `--cpus-per-task` for at least one long-running rule in your Snakefile, then job groups aren't viable unless you keep the number of jobs per group below the number of cores in each compute node (which in my current situation is 28, which isn't a sufficient improvement when you need to submit thousands of a particular short-running rule). In practice, I create a wrapper script that submits thousands of Snakemake jobs as a Slurm array job with each executed via`snakemake --nolock`.

**Comment by pvandyken** [comment]
@jdblischak, let me check to see that I understand by rephrasing in my own words: The goal is to have an arbitrary number of jobs (based on one more rules), each with an arbitrary number of `threads`, and still have it fit on a cluster node with a limit of, say, 28 cores.

If so, then doing something like `snakemake --cluster 'sbatch ...' --jobs XXX --cores 28` should limit the number of cores requested on any given job submission to 28, assuming the slurm profile is set up correctly. For my own case, my cluster has 32 cores per node, so I just have a line in my `config.yaml` file: `cores: 32` so that it gets applied to every job. I know you mentioned trying `--cores` above, so I maybe I misunderstand the problem?

**Comment by jdblischak** [comment]
> should limit the number of cores requested on any given job submission to 28

I tried this, but it doesn't overrule the fact that I configured `--group-components` to submit 100 rules in each job. It still requests 100 threads and fails.

> I know you mentioned trying --cores above, so I maybe I misunderstand the problem?

This is complex, so I put together a reprex below.

What I want to be achieve:

* The 1000 instances of `fastRuleToGroup` are submitted to Slurm in 10 jobs that process 100 rules each. Ideally this would only require 1 core, but at most 28 cores
* The rule `multiThreadedRule` is submitted to Slurm with 8 threads

What currently happens:

* The group jobs for `fastRuleToGroup` all fail because they request 100 threads, which is not available
* The rule `multiThreadedRule` is executed as expected

Thus in practice I'm not able to combine these 2 types of rules in the same pipeline. If I remove `--cpus-per-task={threads}` from `--cluster`, then I can run the group jobs with 1 core, but then I can't run multi-threaded rules that benefit from multiple threads.

Is there a way to include both these types of rules and have the entire pipeline complete in a single invocation of `snakemake`? Here's the reprex:

```
cat > Snakefile << EOF
n = 1000


rule all:
    input:
        expand("output/{rep}.txt", rep=range(1, n + 1)),
        "output/threads.txt",


rule fastRuleToGroup:
    output:
        "output/{file}",
    resources:
        mem_mb=100,
    shell:
        """
        touch {output}
        """


rule multiThreadedRule:
    output:
        "output/threads.txt",
    threads: 8
    shell:
        """
        sleep 10s
        echo {threads} > {output}
        """
EOF

snakemake \
  --cluster "sbatch --mem={resources.mem_mb} --cpus-per-task={threads}" \
  --default-resources mem_mb=2000 \
  --jobs 500 \
  --groups fastRuleToGroup=groupjob --group-components groupjob=100 \
  --cores 28
```

**Comment by pvandyken** [comment]
@jdblischak, I ran the reprex with the latest version of Snakemake (the only thing I changed was to add `--account my_account` to the slurm call), and everything seemed to work correctly: 10 jobs with 28 cores each were scheduled to run `fastRuleToGroup`, and 1 job with 8 cores to run `multiThreadedRule`. If you're not getting this behaviour, there definitely is a problem, but I'm so far not able to reproduce it.

**Comment by jdblischak** [comment]
Well this is embarrassing. I had returned to this issue after some time, and I accidentally activated the wrong conda env. It didn't work with snakemake 7.9.0, but it does with 7.14.1. I apologize for the confusion, @pvandyken, and I super appreciate your patience.

**Comment by pvandyken** [comment]
Haha, no worries, I'm glad we figured it out!

**Comment by jdblischak** [comment]
Based on our discussion, I've updated my job grouping example for my smk-simple-slurm profile so that more users can take advantage of the recent features you added. Thanks again!

https://github.com/jdblischak/smk-simple-slurm/tree/main/examples/job-grouping

**Comment by pvandyken** [comment]
Very good, thanks. If you care to link to it, the documentation discussion on resources and group jobs can be found here https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#resources-and-group-jobs. In particular, the `runtime` resource can be used to automatically calculate how much time will be needed to run the jobs (given the cores and memory constraints).

If you have any suggestions on improving the docs, I'd love to hear them: I'm not super satisfied with the current write-up. I think it focuses too much on the implementation details and not enough on the big picture.

**Comment by moritzschaefer** [comment]
For future googlers: The split-gather feature of snakemake might also be a fit for this request.

**Comment by Tony-Makdissy** [comment]
Hi there,
I am having a problem changing the profile config file to do the grouping.

I tried the grouping functionality which works like magic! I have thousands of files and submitting all of them to SLURM cluster takes too much time for allocating.

when using the following command:
```
snakemake all --groups copy_files=group0 --group-components group0=10 --profile profile
```

and my `profile/config.v8.yaml` file looks like the following:
```
executor: slurm
jobs: 40
default-resources:
  slurm_account: <MY ACCOUNT> 
```

it groups the rules and submit each 10 together and all good. 

but when I tried the following command and profile together it didn't work as I wanted:
command:
```
snakemake all --profile profile
```
profile config:
```
executor: slurm
jobs: 40
default-resources:
  slurm_account: "umisyn"

groups:
  copy_files:
    copy_rule
group-components:
  copy_rule:
    10

```

I believe such grouping is server-depended, if your server allows for more allocations you should group less jobs together and vice versa so I'm really interested in pushing this grouping to the configuration file.

**Comment by pvandyken** [comment]
I believe the syntax should be something like this:

```yaml
executor: slurm
jobs: 40
default-resources:
  slurm_account: "umisyn"

groups:
  - copy_files=copy_rule
group-components:
  - copy_rule=10
```

In general the yaml arguments should be formatted same as in the CLI. Does this work?
